### DISCUSSION TITLE
Can we implement custom peripheral hardware

### DISCUSSION THREAD

**Discussion by rs2xd** [IBM/aihwkit#453] [General]
Hello, I am new to using the aihwkit and wanted to know how one could add custom hardware to the framework. For instance, could i design a custom peripheral that applies a specific control vector the RCU crossbar array, to emulate a specific memory read signal.

**Comment by maljoras** [comment]
Hi @rs2xd,
that is an interesting suggestion. Indeed, that is very well possible. `AnalogTile` encapsulates the crossbar array as well as the close peripheral (for instance an digital multiplication of the ADC output to get a FP number). 

So one could in principle overload the `AnalogTile` (or `InferenceTile`) and add your peripheral operation to the `forward`, `backward` and `update` functions. 

If the memory read is more on the 'logical linear layer' abstraction level, then one could also build a custom `AnalogLinear` layer that adds the peripheral compute. 

If the memory read signal would affect the crossbar output in some way, then it would have to be added to the internal crossbar computation. However, note that ad this stage we are not estimating any latency / energy compute, so it would only make sense to add this memory read as long it would change the output of the crossbar in some way.
